File: apps/web/lib/cdss/rules.ts
// lib/cdss/rules.ts — FASE 3A: motor de reglas clínicas declarativo y auditable.
// Cada regla es IF condición THEN alerta/acción, con severidad, acción recomendada y la
// evidencia (los valores que la dispararon → rule trace). Reglas cuyos datos no existan
// en el EHR del paciente simplemente no disparan.
import type { PatientSnapshot } from "./snapshot"

export const CDSS_THRESHOLDS = {
  glucosaAlta: 130,
  glucosaMuyAlta: 180,
  imcObesidad: 30,
  pasElevada: 130,
  pasCrisis: 160,
  padCrisis: 100,
  riesgoMuyAlto: 0.4,
  riesgoAlto: 0.2,
  diasSinConsulta: 120,
  cesDeterioro: 40,
  glucosaSlopeAlta: 8,
  pesoSlopeAlta: 1,
} as const

export type Severity = "info" | "warning" | "critical"

export interface FiredRule {
  ruleId: string
  name: string
  severity: Severity
  action: string
  message: string
  priorityFlag: boolean
  evidence: Record<string, number | string | boolean>
}

interface RuleDef {
  id: string
  name: string
  severity: Severity
  action: string
  message: string
  priorityFlag?: boolean
  check: (s: PatientSnapshot) => Record<string, number | string | boolean> | null
}

const T = CDSS_THRESHOLDS

const RULES: RuleDef[] = [
  {
    id: "R01", name: "Hiperglucemia + obesidad", severity: "warning", action: "revisar_tratamiento",
    message: "Glucosa e IMC elevados de forma simultánea.",
    check: (s) =>
      s.glucosa != null && s.imc != null && s.glucosa > T.glucosaAlta && s.imc > T.imcObesidad
        ? { glucosa: s.glucosa, imc: s.imc } : null,
  },
  {
    id: "R03", name: "Riesgo muy alto sin consulta reciente", severity: "critical",
    action: "contacto_preventivo", priorityFlag: true,
    message: "Riesgo muy alto y sin consulta reciente: sugerir contacto preventivo.",
    check: (s) =>
      s.riesgo != null && s.riesgo >= T.riesgoMuyAlto && s.diasSinConsulta != null &&
      s.diasSinConsulta > T.diasSinConsulta
        ? { riesgo: s.riesgo, diasSinConsulta: s.diasSinConsulta } : null,
  },
  {
    id: "R04", name: "Glucosa muy alta", severity: "critical", action: "revisar_tratamiento",
    message: "Glucosa en rango muy alto: revisar tratamiento y reconfirmar medición.",
    check: (s) => (s.glucosa != null && s.glucosa >= T.glucosaMuyAlta ? { glucosa: s.glucosa } : null),
  },
  {
    id: "R05", name: "Crisis hipertensiva", severity: "critical", action: "derivar_especialista",
    priorityFlag: true, message: "Presión arterial en rango de crisis.",
    check: (s) =>
      (s.pas != null && s.pas >= T.pasCrisis) || (s.pad != null && s.pad >= T.padCrisis)
        ? { pas: s.pas ?? 0, pad: s.pad ?? 0 } : null,
  },
  {
    id: "R07", name: "Evolución desfavorable", severity: "warning", action: "evaluar_factores",
    priorityFlag: true, message: "Índice de evolución clínica bajo: deterioro de la evolución.",
    check: (s) => (s.ces != null && s.ces < T.cesDeterioro ? { ces: s.ces } : null),
  },
  {
    id: "R08", name: "Tendencia glucémica creciente", severity: "warning", action: "revisar_tratamiento",
    message: "Glucosa con pendiente creciente sostenida.",
    check: (s) => (s.glucosaSlope >= T.glucosaSlopeAlta ? { glucosaSlope: s.glucosaSlope } : null),
  },
  {
    id: "R10", name: "Datos clínicos desactualizados", severity: "info", action: "actualizar_glucosa",
    message: "Sin registros recientes: actualizar glucosa/signos vitales.",
    check: (s) =>
      s.diasSinConsulta != null && s.diasSinConsulta > T.diasSinConsulta
        ? { diasSinConsulta: s.diasSinConsulta } : null,
  },
  {
    id: "R11", name: "Multicomorbilidad con riesgo alto", severity: "warning",
    action: "derivar_especialista", priorityFlag: true,
    message: "Dos o más comorbilidades con riesgo elevado.",
    check: (s) =>
      s.nComorbilidades >= 2 && s.riesgo != null && s.riesgo >= T.riesgoAlto
        ? { comorbilidades: s.comorbilidades.join(", "), riesgo: s.riesgo } : null,
  },
  {
    id: "R12", name: "Aumento de peso sostenido", severity: "warning", action: "evaluar_factores",
    message: "Peso con pendiente creciente relevante.",
    check: (s) => (s.pesoSlope >= T.pesoSlopeAlta ? { pesoSlope: s.pesoSlope } : null),
  },
]

const SEV_ORDER: Record<Severity, number> = { critical: 0, warning: 1, info: 2 }

/** Evalúa todas las reglas sobre el snapshot y devuelve la traza de las que disparan. */
export function evaluateRules(s: PatientSnapshot): FiredRule[] {
  const fired: FiredRule[] = []
  for (const r of RULES) {
    const evidence = r.check(s)
    if (evidence) {
      fired.push({
        ruleId: r.id, name: r.name, severity: r.severity, action: r.action,
        message: r.message, priorityFlag: !!r.priorityFlag, evidence,
      })
    }
  }
  return fired.sort((a, b) => SEV_ORDER[a.severity] - SEV_ORDER[b.severity])
}

/** Catálogo de reglas (sin lambdas) para auditoría/documentación. */
export function ruleCatalog() {
  return RULES.map(({ id, name, severity, action, message, priorityFlag }) => ({
    id, name, severity, action, message, priorityFlag: !!priorityFlag,
  }))
}

File: ml-research/clinical_decision_support/src/predia_cdss/rules.py
"""FASE 3A — Clinical Rule Engine.

Motor de reglas declarativo y AUDITABLE. Cada regla es `IF condición THEN
alerta/acción/prioridad`, con severidad, mensaje, acción recomendada y una
justificación que cita la EVIDENCIA (los valores que la dispararon) → rule trace.
"""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Callable

from . import config

TH = config.TH


@dataclass
class Rule:
    id: str
    name: str
    severity: str          # info | warning | critical
    action: str            # clave del ACTION_CATALOG
    message: str
    # check(s) -> dict de evidencia si dispara, None si no
    check: Callable[[dict], dict | None] = field(repr=False)
    priority_flag: bool = False  # marca seguimiento prioritario


def _r(id, name, severity, action, message, check, priority_flag=False):
    return Rule(id, name, severity, action, message, check, priority_flag)


RULES: list[Rule] = [
    _r("R01", "Hiperglucemia + obesidad", "warning", "revisar_tratamiento",
       "Glucosa e IMC elevados de forma simultánea",
       lambda s: ({"glucosa": s["glucosa_current"], "imc": s["imc_current"]}
                  if s["glucosa_current"] > TH["glucosa_alta"] and s["imc_current"] > TH["imc_obesidad"]
                  else None)),
    _r("R02", "PA elevada persistente (3 consultas)", "warning", "programar_consulta",
       "Presión arterial elevada en las últimas 3 consultas → seguimiento prioritario",
       lambda s: ({"pa_elevated_last3": True, "pas_actual": s["pas_current"]}
                  if s.get("pa_elevated_last3") else None), priority_flag=True),
    _r("R03", "Riesgo muy alto sin consulta reciente", "critical", "contacto_preventivo",
       "Riesgo muy alto y sin consulta reciente → sugerir contacto preventivo",
       lambda s: ({"riesgo": s["riesgo_current"], "dias_sin_consulta": s["days_since_last_consulta"]}
                  if s["riesgo_current"] >= TH["riesgo_muy_alto"]
                  and s["days_since_last_consulta"] > TH["dias_sin_consulta"] else None),
       priority_flag=True),
    _r("R04", "Glucosa muy alta", "critical", "revisar_tratamiento",
       "Glucosa en rango muy alto: revisar tratamiento y reconfirmar medición",
       lambda s: ({"glucosa": s["glucosa_current"]}
                  if s["glucosa_current"] >= TH["glucosa_muy_alta"] else None)),
    _r("R05", "Crisis hipertensiva", "critical", "derivar_especialista",
       "Presión arterial en rango de crisis",
       lambda s: ({"pas": s["pas_current"], "pad": s["pad_current"]}
                  if s["pas_current"] >= TH["pas_crisis"] or s["pad_current"] >= 100 else None),
       priority_flag=True),
    _r("R06", "Baja adherencia con tratamiento activo", "warning", "reforzar_adherencia",
       "Adherencia por debajo del umbral con medicación activa",
       lambda s: ({"adherencia": s["adherencia"], "n_medicacion": s["n_medicacion"]}
                  if s["adherencia"] < TH["adherencia_baja"] and s["n_medicacion"] > 0 else None)),
    _r("R07", "Evolución desfavorable (CES bajo)", "warning", "evaluar_factores",
       "Clinical Evolution Score bajo: deterioro de la evolución",
       lambda s: ({"ces": s["ces"], "banda": s["ces_band"]}
                  if s["ces"] < 40 else None), priority_flag=True),
    _r("R08", "Tendencia glucémica creciente", "warning", "revisar_tratamiento",
       "Glucosa con pendiente creciente sostenida",
       lambda s: ({"glucosa_slope_m": s["glucosa_slope_m"]}
                  if s["glucosa_slope_m"] >= 8 else None)),
    _r("R09", "Sedentarismo", "info", "educacion_habitos",
       "Actividad física por debajo de lo recomendado",
       lambda s: ({"actividad": s["actividad_current"]}
                  if s["actividad_current"] < TH["actividad_baja"] else None)),
    _r("R10", "Datos clínicos desactualizados", "info", "actualizar_glucosa",
       "Sin registros recientes: actualizar glucosa/signos vitales",
       lambda s: ({"dias_sin_consulta": s["days_since_last_consulta"]}
                  if s["days_since_last_consulta"] > TH["dias_sin_consulta"] else None)),
    _r("R11", "Multicomorbilidad con riesgo alto", "warning", "derivar_especialista",
       "Dos o más comorbilidades con riesgo elevado",
       lambda s: ({"comorbilidades": s["comorbilidades"], "riesgo": s["riesgo_current"]}
                  if s["n_comorbilidades"] >= 2 and s["riesgo_current"] >= TH["riesgo_alto"] else None),
       priority_flag=True),
    _r("R12", "Aumento rápido de peso", "warning", "evaluar_factores",
       "Peso con pendiente creciente relevante",
       lambda s: ({"peso_slope_m": s["peso_slope_m"]}
                  if s["peso_slope_m"] >= 1.0 else None)),
]


def evaluate(snapshot: dict) -> list[dict]:
    """Evalúa todas las reglas sobre un snapshot. Devuelve la traza de las que disparan."""
    fired = []
    for rule in RULES:
        ev = rule.check(snapshot)
        if ev is not None:
            fired.append({
                "rule_id": rule.id, "name": rule.name, "severity": rule.severity,
                "action": rule.action, "message": rule.message,
                "priority_flag": rule.priority_flag, "evidence": ev,
            })
    sev_order = {"critical": 0, "warning": 1, "info": 2}
    return sorted(fired, key=lambda f: sev_order[f["severity"]])


def export_rules() -> list[dict]:
    """Catálogo de reglas (sin las lambdas) para auditoría/persistencia."""
    return [{"id": r.id, "name": r.name, "severity": r.severity, "action": r.action,
             "message": r.message, "priority_flag": r.priority_flag} for r in RULES]
